src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -296,29 +535,193 @@ private void PutCounterImpl(DocumentsOperationContext context, string documentId
                         table.Set(tvb);
                     }
+                }
+            }
+        }
 
-                    UpdateMetrics(counterKey, name, changeVector, collection);
+        private static Dictionary<LazyStringValue, int> DbIdToIndex(BlittableJsonReaderArray dbIds)
+        {
+            var count = 0;
+            return new Dictionary<LazyStringValue, int>(dbIds.Items.ToDictionary(x => (LazyStringValue)x, x => count++));
+        }
 
-                    context.Transaction.AddAfterCommitNotification(new CounterChange
+        private long PutNewCounter(DocumentsOperationContext context, BlittableJsonReaderObject data, string documentId, string counterName, 
+            BlittableJsonReaderArray localDbIds, BlittableJsonReaderArray sourceDbIds, Dictionary<LazyStringValue, int> dict, BlittableJsonReaderObject.RawBlob counterValues)
+        {
+            _countersCount++;
+
+            long value = 0;
+            using (GetCounterPartialKey(context, documentId, counterName, out var counterKey))
+            {
+                RemoveTombstoneIfExists(context, counterKey);
+            }
+
+            var sourceCount = counterValues.Length / SizeOfCounterValues;
+            var maxDbIdIndex = FindMaxDbIdIndex(sourceCount, dict, localDbIds, sourceDbIds);
+            var sourceValues = (CounterValues*)counterValues.Ptr;
+
+            _scopes.Add(context.Allocator.Allocate((maxDbIdIndex + 1) * SizeOfCounterValues, out var newVal));
+
+            Memory.Set(newVal.Ptr, 0, (maxDbIdIndex + 1) * SizeOfCounterValues);
+
+            data.Modifications = data.Modifications ?? new DynamicJsonValue(data);
+
+            for (int i = 0; i < sourceCount; i++)
+            {
+                var sourceDbId = (LazyStringValue)sourceDbIds[i];
+                var dbIdIndex = dict[sourceDbId];
+
+                var newEntry = (CounterValues*)newVal.Ptr+ dbIdIndex;
+
+                newEntry->Value = sourceValues[i].Value;
+                newEntry->Etag = sourceValues[i].Etag;
+
+                value += sourceValues[i].Value;
+            }
+
+            data.Modifications[counterName] = new BlittableJsonReaderObject.RawBlob
+            {
+                Length = newVal.Length,
+                Ptr = newVal.Ptr
+            };
+
+            return value;
+        }
+
+        private long PutExistingCounter(DocumentsOperationContext context, BlittableJsonReaderObject data, LazyStringValue counterName,
+            BlittableJsonReaderArray localDbIds,
+            BlittableJsonReaderArray sourceDbIds, Dictionary<LazyStringValue, int> dict, BlittableJsonReaderObject.RawBlob existingCounter,
+            BlittableJsonReaderObject.RawBlob source, out bool modified)
+        {
+            long value = 0;
+            modified = false;
+            var existingCount = existingCounter.Length / SizeOfCounterValues;
+            var sorted = SortValuesByDbIdIndex(dict, localDbIds, sourceDbIds, source.Length / SizeOfCounterValues);
```

Comment:
SortValuesByDbIdIndex's main purpose seems to be  making sure the dict holds all dbIDs, after that, the iteration in line 610 can be performed on the dictionary values, I don't think we need "sorted" at all here, or am I missing something?

Reply:
no, it's not just to make sure the dict holds all dbIDs.
the main purpose here was to avoid calling `context.Allocator.Allocate()` multiple times and moving the data around.
by sorting the source dbIDs by their local dbId-index, then iterating the sorted array in reverse order, we assure that we have a single call to `context.Allocator.Allocate()` (at most) and that we're not moving the data more than once.

Reply:
I don't follow why this is required? But in general we should kill the dictionary and not do any sorting

Reply:
Do this, first read the `localDbIds` to a List, then scan this.
Add to this list if needed with new values.

Same logic, simpler behavior. There is no need to do much here, we expect 3 - 5 max.

Reply:
done